dll/win32/browseui/shellbrowser.cpp

```diff
@@ -3551,6 +3554,156 @@ LRESULT CShellBrowser::RelayMsgToShellView(UINT uMsg, WPARAM wParam, LPARAM lPar
+{
+    INT cItems = ::GetMenuItemCount(hMenu);
+    BOOL bPrevSep = FALSE;
+    for (INT iMenu = cItems - 1; iMenu >= 0; --iMenu) // In reverse order, to keep the indexes
+    {
+        // Get menu type
+        MENUITEMINFOW mii = { sizeof(mii), MIIM_TYPE };
+        GetMenuItemInfoW(hMenu, iMenu, TRUE, &mii);
+        if (mii.fType == MFT_SEPARATOR)
+        {
+            if (bPrevSep)
+                ::DeleteMenu(hMenu, iMenu, MF_BYPOSITION);
+            bPrevSep = TRUE;
+        }
+        else
+        {
+            bPrevSep = FALSE;
+        }
+    }
+}
+
+HRESULT CShellBrowser::ShowSysMenuContextMenu(INT xPos, INT yPos)
+{
+    HRESULT hr;
+    CComPtr<IContextMenu> pContextMenu;
+    CComPtr<IShellFolder> pFolder;
+    LPCITEMIDLIST pidlChild;
+
+    // Prepare for context menu
+    hr = SHBindToParent(fCurrentDirectoryPIDL, IID_PPV_ARG(IShellFolder, &pFolder), &pidlChild);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+    hr = pFolder->GetUIObjectOf(m_hWnd, 1, &pidlChild, IID_IContextMenu, NULL, (void **)&pContextMenu);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    // Query context menu
+#define CMDID_FIRST 0x1000
+#define CMDID_LAST  0x7000
+    UINT uFlags = CMF_NORMAL;
+    HMENU hMenu = ::CreatePopupMenu();
+    hr = pContextMenu->QueryContextMenu(hMenu, 0, CMDID_FIRST, CMDID_LAST, uFlags);
+    if (FAILED_UNEXPECTEDLY(hr))
+    {
+        ::DestroyMenu(hMenu);
+        return hr;
+    }
+
+    // Delete some menu items
+    DeleteUnwantedVerbItems(pContextMenu, hMenu, CMDID_FIRST, CMDID_LAST);
+    DeleteDuplicateSeparators(hMenu);
+    C_ASSERT(!(CMDID_FIRST <= SC_CLOSE && SC_CLOSE <= CMDID_LAST));
+#undef CMDID_FIRST
+#undef CMDID_LAST
+
+    // Get 'Close' menu item string from system menu
+    WCHAR szClose[128];
+    HMENU hSysMenu = ::GetSystemMenu(m_hWnd, FALSE);
+    MENUITEMINFOW mii = { sizeof(mii), MIIM_TYPE };
+    mii.dwTypeData = szClose;
+    mii.cch = _countof(szClose);
+    if (!::GetMenuItemInfoW(hSysMenu, SC_CLOSE, FALSE, &mii))
+        StringCchCopyW(szClose, _countof(szClose), L"&Close\tAlt+F4");
+    // Insert 'Close' menu item and make it default
+    mii.fMask = MIIM_ID | MIIM_FTYPE | MIIM_STATE | MIIM_STRING | MIIM_BITMAP;
+    mii.fType = MFT_STRING;
+    mii.fState = MFS_DEFAULT;
+    mii.wID = SC_CLOSE;
+    mii.dwTypeData = szClose;
+    mii.hbmpItem = HBMMENU_POPUP_CLOSE;
+    ::InsertMenuItemW(hMenu, 0, TRUE, &mii);
+    // Insert a separator
+    mii.fMask = MIIM_FTYPE;
+    mii.fType = MFT_SEPARATOR;
+    ::InsertMenuItemW(hMenu, 1, TRUE, &mii);
+    // No left-side space for checkmarks
+    MENUINFO mi = { sizeof(mi), MIM_STYLE };
+    ::GetMenuInfo(hMenu, &mi);
+    mi.dwStyle |= MNS_NOCHECK;
+    ::SetMenuInfo(hMenu, &mi);
+
+    // Display the menu
+    ::SetForegroundWindow(m_hWnd);
+    UINT id = ::TrackPopupMenu(hMenu, TPM_LEFTALIGN | TPM_RETURNCMD | TPM_RIGHTBUTTON,
+                               xPos, yPos, 0, m_hWnd, NULL);
+    // Action!
```

Comment:
```suggestion
```

Reply:
By the way, why not moving those action things *after* destroying the menu (i.e. after that DestroyMenu call below) ?

Reply:
Fixed.